### ISSUE TITLE
httpsOnly and serviceDomainMatchAll parameters

### ISSUE THREAD

**Issue by mcctomsk** [vfarcic/docker-flow-proxy#248]
I have several services through docker stack. One of them uses this parameters:
```
      labels:
        com.df.notify: "true"
        com.df.distribute: "true"
        com.df.aclName: "95-service-5"
        com.df.serviceDomainMatchAll: "true"
        com.df.servicePath: "/"
        com.df.httpsOnly: "true"
        com.df.serviceDomain: "foo.bar"
        com.df.port: "5000"
```
another one uses this:
```
   labels:
        com.df.notify: "true"
        com.df.distribute: "true"
        com.df.aclName: "40-service-6"
        com.df.servicePath: "/"
        com.df.serviceDomain: "subdomain.foo.bar"
        com.df.port: "5000"
```
In this configuration the subdomain.foo.bar service isn't accessible. When you try to open it via http protocol browser gets a redirection (302 code) to https but service doesn't support it.

Is it possible to fix this?

**Comment by vfarcic** [comment]
Can you please let me know what you'd like to accomplish? Did you put certificates to the proxy and want to allow only HTTPS access to it or are your certificates in the downstream service?

**Comment by mcctomsk** [comment]
It look like we have misunderstanding there. From documentation: "httpsOnly - If set to true, HTTP requests to the service will be redirected to HTTPS." 
This mean that 
1. If Client send request to service -- via http --> but HAproxy tells client it should send him a request to this service via https only.
2. If Client send request to service -- via https --> HAProxy forward it -- via http --> to internal service.

Am I right?

I have a wildcard certificate as a docker secret.
Right now if I set `com.df.httpsOnly: "true"` and `com.df.serviceDomainMatchAll: "true"` for one service DFP is redirecting all http requests to https for all other services. In browser in console log I see this: "Request Method:GET; Status Code:302 Found" and next request via https.
But if I set `com.df.httpsOnly: "true"` only for one service as example "subdomain.foo.bar" then all other services will be available  via http. **And I need vise versa. All service available via https and only one service via http. Is it possible in one configuration?**

**Comment by vfarcic** [comment]
@mcctomsk You're right. There can be other combinations but the one you describes is what's (probably) happening in your case.

`serviceDomainMatchAll` does not match all domains but only those of the specified domain and it's subdomains. For example, if `serviceDomain` is `foo.bar`, it would also match `something.foo.bar` but not `something.acme.com`.

At the moment, you'd would need to remove `serviceDomainMatchAll` to accomplish what you're trying to do. The problem is that you'd need to explicitly specify all the domains and subdomains (if can be all part of the same service) and set the one that should be available through http separately. I guess that's precisely the thing you're trying to avoid.

**Comment by vfarcic** [comment]
Actually, what I wrote in the previous post is not correct. You can do what you need if you make sure that the http service is specified before the others. ACLs are set in alphabetic order using `aclName` parameter. If it is not specified, it defaults to `serviceName`. If for example, you set one label to `com.df.aclName=01-service-with-http` and the other to `com.df.aclName=02-service-with-https-only`, the first one will be evaluated first and, if there is a match, it'll be forwarded to the destination service. As long as only the second one has `httpsOnly`, it should do what you need it to do.

Do you think that would solve your problem?

Keep in mind that I haven't tried this specific scenario so I might be wrong. Can you try it out?

**Comment by mcctomsk** [comment]
>  I guess that's precisely the thing you're trying to avoid.
Exactly. Subdomains will created by user, so I can't write all of them.

> Do you think that would solve your problem?
It will be grate if it works. I've tried to reorder services via aclName but it didn't help.
Рhonestly speaking it's needed for testing stage only. On production stage a service which no need https will be in some private network on another one server.

**Comment by vfarcic** [comment]
@mcctomsk I'll try it myself during the weekend and get back to you. I hope that's not too late.

**Comment by vfarcic** [comment]
Can you send me the proxy config that was generated?

**Comment by mcctomsk** [comment]
Ready. To repeat insert please into /etc/hosts file next rows

> 127.0.0.1 svctest.local
> 127.0.0.1 a.svctest.local

Folder 1 -  `com.df.httpsOnly: "true"` and `com.df.serviceDomainMatchAll: "true"`
Folder 2 - vise versa `com.df.httpsOnly: "true"` has service without `com.df.serviceDomainMatchAll: "true"`

[test.zip](https://github.com/vfarcic/docker-flow-proxy/files/1051686/test.zip)

**Comment by vfarcic** [comment]
Can you send me the configuration generated by the proxy? You can do that by executing:

```bash
docker container exec -it [ID_OF_A_PROXY_CONTAINER] cat /cfg/haproxy.cfg
```

**Comment by mcctomsk** [comment]
haproxy.cfg if a service has `com.df.httpsOnly: "true"` and `com.df.serviceDomainMatchAll: "true"`:

> global
>     pidfile /var/run/haproxy.pid
>     tune.ssl.default-dh-param 2048
> 
>     # disable sslv3, prefer modern ciphers
>     ssl-default-bind-options no-sslv3
>     ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
> 
>     ssl-default-server-options no-sslv3
>     ssl-default-server-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
> 
> resolvers docker
>     nameserver dns 127.0.0.11:53
> 
> defaults
>     mode    http
>     balance roundrobin
> 
>     option  dontlognull
>     option  dontlog-normal
>     option  http-keep-alive
>     option  forwardfor
>     option  redispatch
> 
>     errorfile 400 /errorfiles/400.http
>     errorfile 403 /errorfiles/403.http
>     errorfile 405 /errorfiles/405.http
>     errorfile 408 /errorfiles/408.http
>     errorfile 429 /errorfiles/429.http
>     errorfile 500 /errorfiles/500.http
>     errorfile 502 /errorfiles/502.http
>     errorfile 503 /errorfiles/503.http
>     errorfile 504 /errorfiles/504.http
> 
>     maxconn 5000
>     timeout connect 5s
>     timeout client  20s
>     timeout server  20s
>     timeout queue   30s
>     timeout tunnel  3600s
>     timeout http-request 5s
>     timeout http-keep-alive 15s
> 
> 
> frontend services
>     bind *:80
>     bind *:443
>     mode http
> 
>     acl url_95-service-55000 path_beg /ext/
>     acl domain_95-service-5 hdr(host) -i a.svctest.local
>     use_backend test_svc2-be5000 if url_95-service-55000 domain_95-service-5
>     acl url_99-service-45000 path_beg /
>     acl domain_99-service-4 hdr_dom(host) -i svctest.local
>     redirect scheme https if !{ ssl_fc } url_99-service-45000 domain_99-service-4
>     use_backend test_svc1-be5000 if url_99-service-45000 domain_99-service-4
> 
> 
> 
> 
> 
> 
> 
> backend test_svc2-be5000
>     mode http
>     http-request add-header X-FORWARDED-HOST %[req.hdr(Host)]
>     http-request set-path %[path,regsub(/ext/,/)]
>     server test_svc2 test_svc2:5000
> 
> 
> backend test_svc1-be5000
>     mode http
>     http-response add-header X-XSS-Protection "1; mode=block"
>     http-response add-header X-Frame-Options "DENY"
>     http-response add-header X-Content-Type-Options "nosniff"
>     http-response add-header Content-Security-Policy "script-src self"
>     server test_svc1 test_svc1:5000
> 

haproxy.cfg if a service has `com.df.httpsOnly: "true"` without `com.df.serviceDomainMatchAll: "true"`

> global
>     pidfile /var/run/haproxy.pid
>     tune.ssl.default-dh-param 2048
> 
>     # disable sslv3, prefer modern ciphers
>     ssl-default-bind-options no-sslv3
>     ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
> 
>     ssl-default-server-options no-sslv3
>     ssl-default-server-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
> 
> resolvers docker
>     nameserver dns 127.0.0.11:53
> 
> defaults
>     mode    http
>     balance roundrobin
> 
>     option  dontlognull
>     option  dontlog-normal
>     option  http-keep-alive
>     option  forwardfor
>     option  redispatch
> 
>     errorfile 400 /errorfiles/400.http
>     errorfile 403 /errorfiles/403.http
>     errorfile 405 /errorfiles/405.http
>     errorfile 408 /errorfiles/408.http
>     errorfile 429 /errorfiles/429.http
>     errorfile 500 /errorfiles/500.http
>     errorfile 502 /errorfiles/502.http
>     errorfile 503 /errorfiles/503.http
>     errorfile 504 /errorfiles/504.http
> 
>     maxconn 5000
>     timeout connect 5s
>     timeout client  20s
>     timeout server  20s
>     timeout queue   30s
>     timeout tunnel  3600s
>     timeout http-request 5s
>     timeout http-keep-alive 15s
> 
> 
> frontend services
>     bind *:80
>     bind *:443
>     mode http
> 
>     acl url_95-service-55000 path_beg /ext/
>     acl domain_95-service-5 hdr(host) -i a.svctest.local
>     redirect scheme https if !{ ssl_fc } url_95-service-55000 domain_95-service-5
>     use_backend test_svc2-be5000 if url_95-service-55000 domain_95-service-5
>     acl url_99-service-45000 path_beg /
>     acl domain_99-service-4 hdr_dom(host) -i svctest.local
>     use_backend test_svc1-be5000 if url_99-service-45000 domain_99-service-4
> 
> 
> 
> 
> 
> 
> 
> backend test_svc2-be5000
>     mode http
>     http-request add-header X-FORWARDED-HOST %[req.hdr(Host)]
>     http-request set-path %[path,regsub(/ext/,/)]
>     server test_svc2 test_svc2:5000
> 
> 
> backend test_svc1-be5000
>     mode http
>     http-response add-header X-XSS-Protection "1; mode=block"
>     http-response add-header X-Frame-Options "DENY"
>     http-response add-header X-Content-Type-Options "nosniff"
>     http-response add-header Content-Security-Policy "script-src self"
>     server test_svc1 test_svc1:5000

**Comment by vfarcic** [comment]
`serviceDomainMatchAll=true` messes up the things. I'll do my best to figure out the cause of the problem and fix it ASAP. In the meantime, the workaround is not to use `serviceDomainMatchAll=true`.

**Comment by vfarcic** [comment]
I don't think it will be possible to make it work in that way. Redirection to https is performed independently of the order of ACLs and is performed if there is a match. Since you specified that you want to match any domain or subdomain, it'll redirect to HTTPS no matter what other services specify (unless they are on a different domain). I think that the only way to accomplish what you're trying is to remove `serviceDomainMatchAll=true`. That would result in redirection to https only if a request comes to that specific (sub)domain.

Can you tell me what is your use case and what are you trying to accomplish (in plain words, not labels and configs). I'll do my best to come up with an alternative solution.

**Comment by vfarcic** [comment]
It might not be impossible to do this but I'm not sure how. Can you dig through HAProxy docs? If you can, you can change the `/cfg/haproxy.cfg` file inside a container and send a [reload request](http://proxy.dockerflow.com/usage/#reload). If the changes work, I can easily add them to DFP.

**Comment by vfarcic** [comment]
I think I figured out how to do this after all. Let me give it another try...

**Comment by vfarcic** [comment]
I used the commands that follow to test the fix.

```bash
# Add svctest.local and a.svctest.local to /etc/hosts

docker network create -d overlay proxy

docker stack deploy -c proxy.yml proxy

docker stack deploy -c go-demo.yml go-demo

curl -i "http://localhost:8080/v1/docker-flow-proxy/config"

curl -i "http://localhost:8080/v1/docker-flow-proxy/reload"

curl -i "http://svctest.local/demo/hello" # Redirects to HTTPS

curl -i "http://a.svctest.local/demo/hello" # Does NOT redirect to HTTPS
```

The fix is available in the release 1.408. Can you please try it out and let me know if it works as expected?
